ml_data_management_code.R

```diff
@@ -4,16 +4,266 @@
 ##
 #############################
 
+#
+#
 # Merging data across time points
+#
+#
 
+#
+# Zach Method (built in studying the use case of needing to bind and merge 
+# different datasets that had different elements of missing data in each...
+# this also works to bind datasets together as well)
+#
 
+library(tidyverse)
 
+# create dummy data
+x_pre <- data.frame(p = c(1, 2, 3), a = c(1, 2, 3), b = c(4, 5, 6), d = c(3, 4, 5))
+x_post <- data.frame(p = c(1, 2, 3), a = c(5, 2, 9), c = c(4, 5, 6))
+y_pre <- data.frame(p = c(4, 5, 6), b = c(1, 2, 3), c = c(4, 5, 6))
+y_post <- data.frame(p = c(4, 5, 6), b = c(1, 2, 3), d = c(4, 5, 6), e = c(1, 2, 3))
 
 
+# make function
+my_fun <- function(df){
+  df_name <- deparse(substitute(df))
+  if (grepl("pre", df_name)) {
+    colnames(df)[-1] <- paste(colnames(df)[-1], "t1", sep = "_") 
+  }
+  else if (grepl("post", df_name)) {
+    colnames(df)[-1] <- paste(colnames(df)[-1], "t2", sep = "_")
+  }
+  return(df)
+}
+
+# create a list of data frames
+my_list <- list(x_pre = x_pre, x_post = x_post,
+                y_pre = y_pre, y_post = y_post)
+
+# for loop through list
+for (i in seq_along(my_list)) {
+  df_name <- names(my_list)[i]
+  if (grepl("pre", df_name)) {
+    colnames(my_list[[i]])[-1] <- paste(colnames(my_list[[i]])[-1], "t1", sep = "_") 
+  }
+  else if (grepl("post", df_name)) {
+    colnames(my_list[[i]])[-1] <- paste(colnames(my_list[[i]])[-1], "t2", sep = "_")
+  }
+}
+
+# alternative approach (probably better)
+# uses anonymous function and map
+# re-create the initial list of data frames
+my_list <- list(x_pre = x_pre, x_post = x_post,
+                y_pre = y_pre, y_post = y_post)
+
+my_list <- Map(
+  function(list_df, list_names) {
+    if (endsWith(list_names, "pre")) {
+      names(list_df)[-1] <- paste(names(list_df)[-1], "t1", sep = "_")
+    } 
+    else if (endsWith(list_names, "post")) {
+      names(list_df)[-1] <- paste(names(list_df)[-1], "t2", sep = "_")
+    }
+    list_df
+  }, 
+  my_list, 
+  names(my_list)
+)
+
+#merge all data frames together
+final_product <- my_list %>% 
+  reduce(full_join)
+
+
+#
+# Matt Method (before Zach method)
+#
+
+## not shown here, but typically need to clean dataset before merging
+
+merged_data <- data_pre %>% full_join(data_post, by = c("x"), suffix = c("_pre", "_post"))
```

Comment:
This is comically more simple than the Zach method. Is this approach doing the same thing or are their other steps not being shown?

Reply:
There are other considerations not being shown here. This approach is optimal for merging across columns, not for binding rows. Perhaps you can bind rows with this function and I simply haven't learned how to do that, but I haven't gathered that from what I've done so far.

Reply:
Joins are from SQL, so there is probably a way to do it. Bind_rows and bind_cols are faster, but likely don't matter because we have smaller datasets. Would be good to have different approaches for different circumstances. Table to a future conversation. For MLC, maybe we should just have everything stored in a database like SQL. Options: bind, joins, merge functions. Could also profile code to see difference in speed.